Joons / 태태 Step: 2 로그인 화면 구현

uuu1101 opened the pull request:
- UITextFieldDelegate 의 textFieldShouldReturn 활용하여 ID TextField와 Password TextField에 resignFirstResponder, becomeFirstResponder 메소드를 활용하여 리턴키를 입력받을 시 ID TextField에서는 Password TextField로 Responder를 변경 시키고, Password TextField에서는 Sign In 버튼을 눌렀을 때의 행동을 실행하게 코드를 구성하였습니다.

- idTextFieldConfig, passwordTextFieldConfig 함수를 활용하여 스토리보드로 텍스트필드의 옵션을 수정하면 잘 알아보기 힘들 것 같아 다음 함수에서 코드로 구현하면 다른사람이 봤을 때에도 텍스트필드에 어떤 옵션이 주어졌는지 알아보기 쉬울 것 같아 작성하였습니다.

- touchesBegan, gestureRecognizer의 차이점 
> touchesBegan : 화면에 손만 올려도 터치로 인식  즉, pinch,swipe,drag 동작을 하여도 모두 키보드가 내려가는 현상이 발생
> UITapGestureRecognizer : 화면에 한 손가락 터치 또는 여러 손가락 터치만을 인식
이러한 차이점 때문에 터치에 대한 명확한 보장을 위해서 UITapGestureRecognizer를 사용하였습니다.

elddy0948 committed:
Merge remote-tracking branch 'uuu1101/main' into 1-joons

elddy0948 committed:
add : 스토리 보드에 UI 구성요소 추가

스토리보드에 앱 로고 ImageView, TextField(ID, Password), Button(로그인, 회원가입) 을 추가.

elddy0948 committed:
fix : SignUp 버튼 tint color 수정

Sign Up 버튼의 tint color를 System Red Color로 수정

elddy0948 committed:
add : 스토리보드의 UI구성요소와 ViewController 연결 추가

스토리보드와 ViewController을 IBOutlet과 IBAction을 사용하여 연결.

elddy0948 committed:
feat : TextField의 속성을 정의해 줄 함수 생성	

idTextField와 passwordTextField의 각각의 특성에 맞는 속성을 정의하기 위한 idTextFieldConfig(), passwordTextFieldConfig() 함수 정의.

elddy0948 committed:
feat : UITextFieldDelegate 및 textFieldShouldReturn 메소드 정의 

UITextFieldDelegate를 사용하여 textFieldShouldReturn메소드를 사용하여 idTextField에서 Retrun Key를 입력받을 시 passwordTextField로 이동하게 구현.

elddy0948 committed:
add : Sign In Button, Sign Up Button IBOutlet 생성

Sign In Button, Sign Up Button IBOutlet 생성

elddy0948 committed:
refactor : passwordTextField에서 리턴 키를 입력할 시 기능 추가

passwordTextField에서 리턴키를 입력할 시 didTapSignInButton()가 실행되도록 구현.

elddy0948 committed:
feat : TapGesture 추가

TapGesture에 IBAction을 연결하고 view.endEditing을 사용하여 빈화면을 탭 할 시 키보드가 내려가게 구현

elddy0948 committed:
add : RegisterStepOneViewController 추가

RegisterStepOneViewController추가

elddy0948 committed:
Merge branch 'joons/STEP2' into 1-joons

elddy0948 committed:
fix : 아무 기능을 넣지않은 IBAction 삭제

Sign In Button과 Sign Up Button의 IBAction 삭제

elddy0948 committed:
Merge branch 'joons/STEP2' into 1-joons

uuu1101 committed:
Merge pull request #7 from elddy0948/1-joons

Joons Pull Request

uuu1101 committed:
refactor: textFieldShouldRetrun()

해당 뷰에서 더 이상 textField 가 추가 되지 않을거라고 판단하여
switch 문을 if else 문으로 수정

uuu1101 committed:
fix: main.stroyboard

삭제된 @IBAction func()의 Received Actions 연결 해제

Limwin94 commented on SignUpFlow/SignUpFlow/ViewController.swift:
```diff
@@ -1,18 +1,49 @@
+        textField.autocorrectionType = .no
+        textField.returnKeyType = .next
+    }
+    private func passwordTextFieldConfig(_ textField: UITextField) {
```
```swift
private func configureIDTextField() {
    self.idTextField.autocapitalizationType = .non
    ....
}
```
이렇게 직접 호출해서 변경하는 것과 어떤 차이가 있을까요?

elddy0948 replied:
직접 호출해서 변경하는 것은 idTextField에 직접 접근하여 그 TextField의 속성을 바꿔주는 반면에, UITextField 타입을 가지는 파라미터를 통해 변경해 주면 파라미터를 통해 전달받은 idTextField를 복사한 것을 textField라는 이름으로 새롭게 생성하여 textField의 속성을 변경한 후 함수가 종료될 때 idTextField로 textField의 속성을 넘겨주는 것으로 알고있습니다.
그리고, 나중에 혹시라도 TextField의 이름이 변경되어야 할 경우가 생긴다면 좀 더 유연하게 대처하기 위해 파라미터로 TextField를 받는 방법을 선택하였습니다. 😄

yagom replied:
그렇다면 메서드 이름을 좀 더 범용적으로 사용할 수 있도록 변경해야하지 않을까요?
password 텍스트 필드에만 적용할 수 있는 메서드라면 굳이 매개변수로 텍스트필드를 받아올 필요성을 못느끼겠습니다.

Limwin94 commented on SignUpFlow/SignUpFlow/ViewController.swift:
```diff
@@ -1,18 +1,49 @@
+    
+    //MARK: - TextField Configuration
+    private func idTextFieldConfig(_ textField: UITextField) {
+        textField.autocapitalizationType = .none
```
스토리보드에서도 설정할 수 있는 값을 명시적으로 설정해준 이유가 따로 있을까요? 🧐

elddy0948 replied:
물론 스토리보드에서도 설정할 수 있지만, 코드로 작성해 두면 다른 사람이 봤을때 TextField에 어떠한 옵션들이 들어갔는지 좀 더 확인하기 쉬울거라 생각하여 코드로 작성해 두었습니다! 😊

Limwin94 commented on SignUpFlow/SignUpFlow/ViewController.swift:
```diff
@@ -1,18 +1,49 @@
+    
+    //MARK: - IBOutlets
+    @IBOutlet weak var logoImageView: UIImageView!
+    @IBOutlet weak var idTextField: UITextField!
```
키보드 타입이 Default일때와 Email-Address일때 무슨 차이가 있는걸까요?

elddy0948 replied:
Email-Address 키보드 타입일 때는 Default타입과 다르게 이메일을 작성하는데에 필요한 '@'나 '.' 과 같은 문자열이 키보드에 포함되어 보다 쉽게 작성할 수 있도록 돕습니다. 하지만 이번 프로젝트에서는 email이 아닌 id를 입력받는 TextField라 생각하여 Default타입으로 그냥 두었습니다. 🙌

uuu1101 replied:
프로젝트 명세서상 아이디 형식이 email 이라서 email 타입으로 변경해야 할 것 같습니다 !

Limwin94 reviewed (commented):
고생하셨습니다 👍🏻 

코멘트 남겨보았으니 확인 부탁드립니다~!

추가로 다음화면인 [회원가입 과정 1]을 모달로 올려주게 되면 그 다음화면(과정 2)으로 넘어갈때 또 다시 모달로 띄워야 할까요? 화면흐름이 어색해질수도 있을것 같다고 생각해요.

elddy0948 committed:
Merge remote-tracking branch 'uuu1101/main' into 1-joons

elddy0948 committed:
refactor : idTextFieldConfig() 내의 keyboardType 변경

id TextField의 인풋 값이 email address이므로 keyboardType을 적절하게 설정.

elddy0948 committed:
refactor : 과정 1 화면으로 넘어가는 부분에 Navigation Controller 추가

elddy0948 committed:
Merge branch 'joons/STEP2-1' into 1-joons

elddy0948 committed:
add : 스토리 보드에 UI 추가 및 ViewController와 연결

스토리 보드에 UI 추가 및 ViewController와 연결

elddy0948 committed:
fix : 주석 추가

주석 추가

elddy0948 committed:
feat : profile image view를 탭할 시 photo library로 이동

profile image view를 탭하는 것을 인지하는 UITapGestureRecognizer를 만들고, UIImagePicker를 활용하여 이미지를 선택할 수 있게 함.

elddy0948 committed:
feat : 취소버튼 누를 시 모든 칸을 비우고 로그인 화면으로 돌아감

취소버튼을 누를 시 모든 칸을 nil로 만들고 dismiss 실행

elddy0948 committed:
feat : 모든 Field가 비어있는지에 따라 Next Button 활성화 

NotificationCenter와 프로퍼티 감시자를 활용하여 필드가 비어있는지 비어있지 않은지 확인하고자 하였고, 비어있다면 NextButton 비활성화, 다 채워져있다면 활성화 하게 구현.

elddy0948 committed:
Merge branch 'joons/Step3' into 1-joons

uuu1101 committed:
Merge pull request #8 from elddy0948/1-joons

Pull Request From Joons

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
 import UIKit
 
 class RegisterStepOneViewController: UIViewController {
```
직관적인 이름 좋네요!

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+    @IBOutlet weak var profileImageView: UIImageView!
+    @IBOutlet weak var idTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+    @IBOutlet weak var checkPasswordTextField: UITextField!
```
check의 위치에 따라 동사가될지 아닐지가 결정될 것 같아요.
앞에 있다면 확인한다는 동사일 것 같은데, 그렇다면 이것은 메서드 이름으로 적합하지 않을까요? :)

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+    @IBOutlet weak var idTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+    @IBOutlet weak var checkPasswordTextField: UITextField!
+    @IBOutlet weak var introduceTextView: UITextView!
```
introduce는 동사인 것 같아요 :)

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+        imagePicker.sourceType = .photoLibrary
+        present(imagePicker, animated: true, completion: nil)
+    }
+    @objc func checkAllField() {
```
'무엇'을 확인하는 걸까요?

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+        imagePicker.allowsEditing = true
+        imagePicker.sourceType = .photoLibrary
+        present(imagePicker, animated: true, completion: nil)
+    }
```
아래 한 줄을 띄우지 않은 이유가 있을까요?

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+
+extension RegisterStepOneViewController {
+    @objc func didTapProfileImageView(_ sender: UITapGestureRecognizer) {
+        let imagePicker = UIImagePickerController()
```
이미지뷰를 탭할 때 마다 이미지 피커 컨트롤러를 생성해주는 것과, 뷰 컨트롤러에서 이미지 피커 컨트롤러 인스턴스를 프로퍼티로 갖는 것과 어떤 점에서 차이가 날까요?
코드에서의 가독성, 사용자의 사용성과 성능면에선 어떤 차이를 보일까요?

elddy0948 replied:
뷰 컨트롤러에 인스턴스 프로퍼티를 생성하여 갖고 있으면 힙 영역에 인스턴스가 생성되어 계속 남아있으면서 그것에 대한 속성이 변경될 때에는 그 주소값을 찾아 들어가 속성을 변경해주는 반면에, 이미지 뷰를 탭할 때 마다 생성하는 인스턴스는 탭 할때 부르는 메소드가 끝나면 메모리에서 사라지게 됩니다. 
이렇게 코드를 작성한 이유는 UIImagePickerController는 didTapProfileImageView라는 메소드를 실행할 때에만 사용하는 것이고, 여러군데에서 접근하지 않으며, 자주 접근하지 않을 것 같아서 호출할때에만 만들고 메모리에서 없애버리는 것이 좋다고 생각했습니다. 
그렇기 때문에 위와 같이 ProfileImageView를 탭 할때만 인스턴스를 생성하게 작성하였습니다. 😀

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+    
+    private func checkValidatePassword() -> Bool {
+        //passwordTextField가 비어있지 않고 서로 같으면
+        if (passwordTextField.text != "") && (passwordTextField.text == checkPasswordTextField.text) {
```
텍스트필드의 `text` 프로퍼티가 nil이고, 두 텍스트필드의 `text` 프로퍼티가 모두 nil이면 이 조건식은 안전할까요?
또, 조건식에서 `&&`와 `,`의 차이는 무엇일까요?

elddy0948 replied:
UITextField.text의 타입은 String? 타입으로 nil값이 들어올 수 있습니다. 하지만 IBOutlet을 연결해 주면서 텍스트필드의 값이 ""가 들어가는 것으로 알고있습니다. 하지만 결국 String?타입이고, nil값이 들어올 수 있는 가능성이 있기 때문에 다음 커밋에서 hasText를 활용하여 TextField에 text가 있는지 없는지를 확인 후 패스워드를 비교하는 코드를 구성하였습니다.
조건식에서 &&와 , 는 우선 비슷한 역할을 하는 것으로 알고있습니다. 하지만 , 같은 경우에는 옵셔널 바인딩, 또는 작은 여러개의 condition들을 표현한 condition-list를 나타낼 때 사용하는 것이 좋고, &&는 하나의 큰 condition을 비교할 때 사용하는것이 좋다고 생각하여 사용하게 되었습니다😀

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+            return true
+        }
+        return false
+    }
```
아래 한 줄을 띄우지 않은 이유가 있을까요?

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+    }
+    private func configureTextField(_ textField: UITextField) {
+        if textField == idTextField {
+            textField.autocorrectionType = .no
+            textField.autocapitalizationType = .none
```
모든 경우의 필드에서 모두 공통적으로 사용하는 코드인데 메서드 맨 위쪽으로 공통으로 빼거나, 새로운 메서드로 빼는 것은 어떻게 생각하나요?

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        picker.dismiss(animated: true, completion: nil)
+        guard let pickedImage = info[.editedImage] as? UIImage else {
+            fatalError("No Image Picked!!")
```
fatalError를 발생시켜 사용자의 앱이 튕길만큼 중요한 오류인가요?

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+    @IBAction func didTapNextButton(_ sender: UIButton) {
+        guard let id = idTextField.text, let password = passwordTextField.text, let profileImage = profileImageView.image,
+              let introduce = introduceTextView.text else {
+            fatalError("빈칸이 있습니다!")
```
fatalError를 사용하면 사용자의 앱이 튕깁니다. 그정도로 중요한 문제일까요?

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+            member.members[id]?.setProfileImage(profileImage)
+            member.members[id]?.setIntroduction(introduce)
+        case .failure(let message):
+            fatalError(message.errorDescription ?? "에러 메시지가 없습니다.")
```
fatalError를 사용하면 사용자의 앱이 튕깁니다. 그정도로 중요한 문제일까요?

yagom commented on SignUpFlow/SignUpFlow/RegisterStepOneViewController.swift:
```diff
@@ -8,8 +8,113 @@
+        case .failure(let message):
+            fatalError(message.errorDescription ?? "에러 메시지가 없습니다.")
+        }
+        NotificationCenter.default.removeObserver(UITextField.textDidChangeNotification)
+        NotificationCenter.default.removeObserver(UITextView.textDidChangeNotification)
```
옵저버를 지워주는 코드가 여기보다 더 좋은 위치가 있을 것 같아요.
만약 두 번째 화면에서 되돌아 왔다가 다시 두 번째 화면으로 간다고 하면 어떻게 될까요?
그렇다면 옵저버를 등록하는 코드의 위치도 변경이 되어야 할까요?

yagom reviewed (commented):
여러 고민이 많이 엿보이는 코드였어요 :)
몇 가지 의견 남겨두었습니다.
고민할 포인트가 몇 가지 있어요.
